File: 卡码/代码随想录/1/03-哈希表/Java/src/_05_四数相加/LeetCode_0454.java
//454. 四数相加 II
package _05_四数相加;

import java.util.HashMap;

public class LeetCode_0454 {

}
class Solution {
  public int fourSumCount(int[] nums1, int[] nums2, int[] nums3, int[] nums4) {
  	HashMap<Integer, Integer> map = new HashMap<Integer,Integer>();
  	for(int num1:nums1)
  		for(int num2:nums2) {
  			int target=num1+num2;
  			if(map.containsKey(target))
  				map.put(target,map.get(target)+1);
  			else
					map.put(target,1);
  		}
  	int sum=0;
  	for(int num3:nums3)
  		for(int num4:nums4)
  			if(map.containsKey(-(num3+num4))) sum+=map.get(-(num3+num4));
  	return sum;
  }
}

File: 卡码/代码随想录/1/03-哈希表/Go/05-四数相加/LeetCode_0454.go
//454. 四数相加 II
func fourSumCount(nums1 []int, nums2 []int, nums3 []int, nums4 []int) int {
	sum:=0
	hash:=make(map[int]int)
	for _,num1:=range nums1 {
		for _,num2:=range nums2 {
			hash[num1+num2]++
		}
	}
	for _,num3:=range nums3 {
		for _,num4:=range nums4 {
			target:=-(num3+num4)
			if value,ok:=hash[target];ok {
				sum+=value
			}
		}
	}
	return sum
}
